polynote-kernel/src/main/scala/polynote/kernel/interpreter/python/PythonInterpreter.scala

```diff
@@ -215,6 +228,58 @@ class PythonInterpreter private[python] (
     }
   }
 
+  private def definitionLocations(script: PyObject, line: Int, col: Int) = jep { jep =>
+    val pyNames = script.getAttr("goto", classOf[PyCallable]) // see https://jedi.readthedocs.io/en/latest/docs/api.html#jedi.Script.goto
+      .callAs(
+        classOf[Array[PyObject]],
+        Array[Object](Integer.valueOf(line), Integer.valueOf(col)), // positional args
+        Map[String, Object]("follow_imports" -> Boolean.box(true), "follow_builtin_imports" -> Boolean.box(true)).asJava)
+    pyNames.toList.flatMap {
+      jediName =>
+        val path = jediName.getAttr("module_path", classOf[String])
+        val line = jediName.getAttr("line", classOf[Integer])
+        val column = jediName.getAttr("column", classOf[Integer])
+        if (path != null)
+          Some(DefinitionLocation(s"python://$path", line, column))
+        else None
+    }
+  }
+
+  private def definitionSource(locations: List[DefinitionLocation]) = {
+    for {
+      firstLoc  <- ZIO(locations.headOption).someOrFailException
+      path      <- ZIO(Paths.get(firstLoc.uri.stripPrefix("python://")))
+      source    <- readFile(path)
+    } yield source
+  }.option
+
+  // TODO: we probably need to write out prior code cells, and use the jedi Project (https://jedi.readthedocs.io/en/latest/docs/api.html#jedi.Project)
+  //        to allow finding references to other cells. This will only find references that are external.
```

Comment:
Huh, I'm a little surprised that this only finds external references. I played around a bit: 

```python
In [15]: x = "hi"

In [16]: script = jedi.Interpreter("x", [locals()])

In [17]: script.goto()
Out[17]: [<Name full_name='jedi.api.interpreter.__main__.x', description='instance x'>]

In [18]: script.goto(line=1, column=1)
Out[18]: [<Name full_name='jedi.api.interpreter.__main__.x', description='instance x'>]

In [19]: script.goto(follow_imports=True, follow_builtin_imports=True)
Out[19]: [<Name full_name='jedi.api.interpreter.__main__.x', description='instance x'>]

In [20]: script.goto(line=1, column=1, follow_imports=True, follow_builtin_imports=True)
Out[20]: [<Name full_name='jedi.api.interpreter.__main__.x', description='instance x'>]

In [21]: script.goto(line=1, column=0, follow_imports=True, follow_builtin_imports=True)
Out[21]: [<Name full_name='jedi.api.interpreter.__main__.x', description='instance x'>]
```

Of course the problem is that `jedi.api.interpreter.__main__.x` isn't that useful to us. I guess we could try walking the `State` looking for `x` (i.e., implement finding the definition ourselves if it's a notebook-defined symbol which sounds a little icky but in practice might not be so bad because we are already keeping track of everything?)

Reply:
yeah, I think we will probably have to do what you're suggesting – if jedi fails to find a definition, we can look at the AST, see if there's a variable involved that we have in the `State`, and if so, implement our own jump-to-definition logic (maybe for a direct click on the variable we can just return the source cell, and for a property or method we can look up the type and delegate back to jedi with some dummy code or something). Anyway, will work on that in subsequent PRs.